### PR TITLE
QOL-8179 make objects public when dataset is public

### PR DESCRIPTION
**Pull request by ThrawnCA**

### CODE REVIEW

**Review by duttonw — CHANGES_REQUESTED** [review]
need some tests :)

File: ckanext/s3filestore/uploader.py [hunk 1]
```diff
@@ -496,14 +515,51 @@ def get_path(self, id, filename=None):
...
     def upload(self, id, max_size=10):
         '''Upload the file to S3.'''
 
         # If a filename has been provided (a file is being uploaded) write the
         # file to the appropriate key in the AWS bucket.
         if self.filename:
             filepath = self.get_path(id, self.filename)
-            self.upload_to_key(filepath, self.upload_file)
+            self.upload_to_key(filepath, self.upload_file, acl=self._get_target_acl(id))
+        elif self.acl == 'auto':
```
[the remark this answers is not in the document]
Comment (ThrawnCA) [reply ckanext/s3filestore/uploader.py:561]: `filename` will be absent if we are updating the resource metadata without uploading a new file.

File: ckanext/s3filestore/uploader.py [hunk 2]
```diff
@@ -188,19 +195,28 @@ def get_signed_url_to_key(self, key, extra_params={}):
...
+            if cache_url and is_public_read != _is_presigned_url(cache_url):
                 log.debug('Returning cached URL for path %s', key)
                 return cache_url
             else:
                 log.debug('No cache found for %s; generating a new URL', key)
 
         params = {'Bucket': self.bucket_name,
                   'Key': key}
-        if metadata['ContentType'] != 'application/pdf':
+        if not is_public_read and metadata['ContentType'] != 'application/pdf':
```
[the remark this answers is not in the document]
Comment (ThrawnCA) [reply ckanext/s3filestore/uploader.py:219]: There's no point in adding it to the parameters being used to generate a presigned URL, because we're about to strip off the query string anyway and throw away all the presigning.

(It's actually not a very efficient way to generate the URL, but there [doesn't seem to be direct Boto support for getting the public URL of an object](https://stackoverflow.com/questions/48608570/python-3-boto-3-aws-s3-get-object-url).)

File: ckanext/s3filestore/uploader.py [hunk 3]
```diff
@@ -188,19 +195,28 @@ def get_signed_url_to_key(self, key, extra_params={}):
         # check whether the object exists in S3
         metadata = client.head_object(Bucket=self.bucket_name, Key=key)
 
+        # check whether the object is publicly readable
+        acl = client.get_object_acl(Bucket=self.bucket_name, Key=key)
+        is_public_read = any(
+            grant['Grantee']['Type'] == 'Group'
+            and grant['Grantee'].get('URI', '').endswith('AllUsers')
```
Comment (duttonw) [inline ckanext/s3filestore/uploader.py:202]: unsure if AllUsers is always at end of acl list, should this be contains? 
Also we may want to check for the super override for public visability on the bucket level (outside of acl)
This should be cached for 5min intervals (at least) to reduce api calls (IgnorePublicAcls, RestrictPublicBuckets, s3:GetBucketPublicAccessBlock, s3:GetAccountPublicAccessBlock)
https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-control-block-public-access.html

**Review by duttonw — CHANGES_REQUESTED** [review]
We are also going to need a helper function to diff the acl on the object the privacy setting in the database and if different either:
create an update record for a script to be run at a later date offline 
or
call s3 acl update to fix the permission (make public or remove public)
bit like how you got the file transfer from on disk to s3 working.

File: ckanext/s3filestore/uploader.py (hunk 3 above)
Comment (ThrawnCA) [reply ckanext/s3filestore/uploader.py:202]: As far as I've seen, the 'public-read' grant always uses `http://acs.amazonaws.com/groups/global/AllUsers`; that's what the console interface provides, and that's what you get if you specify the 'public-read' ACL when creating an object.

Comment (duttonw) [reply ckanext/s3filestore/uploader.py:202]: in dev turn on full block via aws console and see what happens on acl calls, (if public read is still visible). My guess is that acl will still turn back what is set, but when accessing file, it will be blocked.

**Comment by ThrawnCA** [comment]
> We are also going to need a helper function to diff the acl on the object the privacy setting in the database and if different either:
> create an update record for a script to be run at a later date offline
> or
> call s3 acl update to fix the permission (make public or remove public)
> bit like how you got the file transfer from on disk to s3 working.

I would like to avoid running a background job, since that would then require Publications to run a batch layer, increasing costs.

We could probably do something with the IPackageController `after_update` hook, though?

File: ckanext/s3filestore/tests/test_uploader.py [hunk 4]
```diff
@@ -217,3 +218,46 @@ def test_resource_upload_with_url_and_clear(self):
...
                        'url': 'http://asdf', 'save': 'save'},
                  extra_environ=env)
+
+    def test_is_presigned_url(self):
+        ''' Tests that presigned URLs are correctly recognised.'''
+        assert_true(_is_presigned_url('https://example.s3.amazonaws.com/resources/foo?AWSAccessKeyId=SomeKey&Expires=9999999999Signature=hb7%2F%2Bz1H%2B8wdEy0pCsX7bZG%2BuPU%3D'))
+        assert_false(_is_presigned_url('https://example.s3.amazonaws.com/resources/foo'))
+
+    @helpers.change_config('ckanext.s3filestore.acl', 'auto')
+    def test_resource_url(self):
```
Comment (duttonw) [inline ckanext/s3filestore/tests/test_uploader.py:228]: please split this into at least 3 tests
test_resource_url_should_be_public_for_publicDataset
test_resource_url_should_be_private_for_privateDataset
test_resource_url_should_update_state_if_private_and_dataset_made_public
test_resource_url_should-update_state_if_public_and_dataset_made_private

also if not already exisitng
test file delete is called

unsure if we should cover this edge case, if filename is renamed, delete old file when uploading new file (after successful upload? )

**Review by duttonw — COMMENTED** [review]
overall it seems ok.

we do need a helper out-of-band script to update all s3 objects to public or private acl, so that we can reduce the number of presigned urls we are generating.

we may want to 'hide' the old file (or delete it) once a successful new file is uploaded if filename is different.

also more tests ;)

File: ckanext/s3filestore/tests/test_uploader.py (hunk 4 above)
Comment (ThrawnCA) [reply ckanext/s3filestore/tests/test_uploader.py]: I can do this, but `test_resource_url_should-update_state_if_public_and_dataset_made_private` seems like it will necessarily be a superset of `test_resource_url_should_be_public_for_publicDataset` (since it will first create a public dataset and resource).

Comment (ThrawnCA) [reply ckanext/s3filestore/tests/test_uploader.py]: The first four suggested tests now exist.

Testing for file deletion on rename may be possible if the test can directly interact with the S3 mock, doing HEAD requests to check whether each file exists.

Comment (ThrawnCA) [reply ckanext/s3filestore/tests/test_uploader.py]: The file deletion testing wasn't working - Moto was reporting 404 whenever I tried to access it directly, not sure why - so I've dropped those for now.

**Review by duttonw — APPROVED** [review]
lets try it in dev, but do have commit id before it stored just in case we need to roll back to do more 'fixes'

**Comment by ThrawnCA** [comment]
> lets try it in dev, but do have commit id before it stored just in case we need to roll back to do more 'fixes'

We can easily roll back via Bamboo. It will auto-deploy as soon as we update the `develop` branch again, though.

**Comment by ThrawnCA** [comment]
Looks like we need to find some way to preserve the `Content-Disposition` headers when downloading directly from S3.

**Comment by duttonw** [comment]
try setting it as metadata on the s3 object (like all other details i've suggested, ie dataset name, org etc)

**Comment by duttonw** [comment]
see https://stackoverflow.com/a/19011398/3217974 for console way, but if set when uploaded, it will have it set correctly.

**Comment by duttonw** [comment]
https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html

**Comment by duttonw** [comment]
http://iwantmyreal.name/s3-download-only-presigned-upload

**Comment by ThrawnCA** [comment]
This is working on DEV now. @duttonw Can you review the last few commits?

File: ckanext/s3filestore/uploader.py [hunk 5]
```diff
@@ -152,12 +189,17 @@ def upload_to_key(self, filepath, upload_file, make_public=False):
...
         try:
-            self.get_s3_resource().Object(self.bucket_name, filepath).put(
-                Body=upload_file.read(), ACL='public-read' if make_public else self.acl,
-                ContentType=mime_type)
+            kwargs = {
+                'Body': upload_file.read(), 'ACL': acl,
+                'ContentType': mime_type}
+            if mime_type != 'application/pdf':
+                filename = filepath.split('/')[-1]
+                kwargs['ContentDisposition'] = 'attachment; filename=' + filename
```
Comment (duttonw) [inline ckanext/s3filestore/uploader.py:197]: can you add more metadata please, department name, dataset uuid, dataset nice name, last uploader username, or just the entire metadata we have if possible (no need to restrict)

Comment (ThrawnCA) [reply ckanext/s3filestore/uploader.py:197]: I'll see what's easily available.

Comment (ThrawnCA) [reply ckanext/s3filestore/uploader.py:197]: The package ID is easy enough to add. However, anything else from the package could become outdated if the package changes.

Comment (duttonw) [reply ckanext/s3filestore/uploader.py:197]: maybe add package_last_update_date to the metadata so that it does give when it was last updated.

Comment (duttonw) [reply ckanext/s3filestore/uploader.py:197]: looking at peoples answers on this and reading the api spec, metadata is set at time of upload and can't be altered until a new file is uploaded or a copyfrom command is issued.  

Lets run this past Mark and co on what fields we should drop. 

https://stackoverflow.com/questions/39596987/how-to-update-metadata-of-an-existing-object-in-aws-s3-using-python-boto3

**Comment by duttonw** [comment]
could you leave these in?

x-amz-meta-package_creator_user_id:ffdfe301-1077-4b03-974e-189eff7ae288
x-amz-meta-package_data_driven_application:NO
x-amz-meta-package_id:5d46d6c5-c083-4593-8d22-2b08466615c7
x-amz-meta-package_license_id:cc-by-4
x-amz-meta-package_license_title:Creative Commons Attribution 4.0
x-amz-meta-package_license_url:https://creativecommons.org/licenses/by/4.0/
x-amz-meta-package_name:testing-qol-8179
x-amz-meta-package_owner_org:579ded79-c901-479f-bb67-043e4437910d
x-amz-meta-package_security_classification:PUBLIC
x-amz-meta-package_state:active
x-amz-meta-package_title:Testing QOL-8179
x-amz-meta-package_type:dataset
x-amz-meta-package_update_frequency:near-realtime

yes some of the nice names will be out of sync but does give good hint on the what the file is to macie and outselves looking at s3.

**Comment by ThrawnCA** [comment]
> could you leave these in?

Okay, done.
